### PR TITLE
Feat/airgap sdr

### PR DESCRIPTION
**Pull request by mdrahman-suse**
#### Proposed Changes ####

Implement airgap cluster tests with system default registry
TO BE NOTED: This is only for Ubuntu as there is a step to `update-ca-certs` which is coded for Ubuntu only. Supporting other OS's will be done in a later PR

Docs updated

#### Types of Changes ####

Implement airgap cluster creation and tests by creating the cluster using system default registry method

#### Testing ####

Checklist:
1. If your PR changes anything on or related to Jenkins, run it pointing to your branch to make sure it's okay.

2. Verify code lint; we should not have errors.

3. Update the documentation if needed.

4. Update makefile and docker run if adding new tests.

5. Run your tests at least 4 times with all configurations needed and possible.

6. If needed test with different os types.

### CODE REVIEW

File: entrypoint/airgap/systemdefaultregistry_test.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+			cluster,
+			assert.PodAssertRestart(),
+			assert.PodAssertReady())
+	})
+
+	AfterAll(func() {
+		shared.DisplayAirgapClusterDetails(cluster)
+	})
+
+	// TODO: Validate deployment, eg: cluster-ip
```
Comment (fmoral2) [inline entrypoint/airgap/systemdefaultregistry_test.go:43]: this will be another PR?

Comment (mdrahman-suse) [reply entrypoint/airgap/systemdefaultregistry_test.go:43]: Yes once we figure out how to do it

Comment (fmoral2) [reply entrypoint/airgap/systemdefaultregistry_test.go:43]: got it ok...will finish review tm

File: entrypoint/airgap/airgap_suite_test.go [hunk 2]
```diff
@@ -75,8 +75,8 @@ func validateAirgap() {
 	}
 }
 
-func TestAirgapSuite(t *testing.T) {
-	RegisterFailHandler(FailWithReport)
+func TestAirgapClusterSuite(t *testing.T) {
+	RegisterFailHandler(Fail)
```
Comment (fmoral2) [inline entrypoint/airgap/airgap_suite_test.go:79]: RegisterFailHandler(FailWithReport)

File: entrypoint/airgap/systemdefaultregistry_test.go (hunk 1 above)
Comment (fmoral2) [reply entrypoint/airgap/systemdefaultregistry_test.go:43]: what exactly u want to validate here?  cause my current PR maybe help you on that

File: modules/airgap/setup/get_artifacts.sh [hunk 3]
```diff
@@ -39,21 +39,29 @@ get_assets() {
   elif [[ "$product" == "rke2" ]]; then
     url="https://github.com/rancher/rke2/releases/download/$version"
     wget "$url"/sha256sum-"$arch".txt
-    wget "$url"/rke2-images.linux-"$arch".txt
+    # Ref: https://docs.rke2.io/install/airgap
+    if [[ -n "$flags" ]] && [[ "$flags" =~ "cni" ]]; then
```
Comment (fmoral2) [inline modules/airgap/setup/get_artifacts.sh:43]: sorry missed that earlier....can we have a doc here at the begining showing example of what are or could be those flags param?

File: modules/airgap/setup/get_artifacts.sh [hunk 4]
```diff
@@ -4,14 +4,14 @@
...
-product=$1
-version=$2
-arch=$3
+product=${1}
+version=${2}
+arch=${3}
 flags=${4}
-tarball_type=$5
+tarball_type=${5}
 prodbin=$product
```
Comment (fmoral2) [inline modules/airgap/setup/get_artifacts.sh:15]: binary? binary_name? 

prodbin seems odd for me. 

*nit pick

File: modules/airgap/setup/get_artifacts.sh [hunk 5]
```diff
@@ -4,14 +4,14 @@
 # set -x
 # echo "$@"
 
-# Usage: ./get_artifacts.sh k3s v1.27.5+k3s1
-# Usage: ./get_artifacts.sh rke2 v1.27.5+rke2r1 amd64 "flags" tar.gz
+# Usage: ./get_artifacts.sh "k3s" "v1.31.0+k3s1"
+# Usage: ./get_artifacts.sh "rke2" "v1.31.0+rke2r1" "amd64" "flags" "tar.gz"
```
Comment (fmoral2) [inline modules/airgap/setup/get_artifacts.sh:8]: can we add basic validations here?
for mandatory vars that cannot get to here empty, like prod and version

File: modules/airgap/setup/get_artifacts.sh [hunk 6]
```diff
@@ -39,21 +39,29 @@ get_assets() {
   elif [[ "$product" == "rke2" ]]; then
     url="https://github.com/rancher/rke2/releases/download/$version"
     wget "$url"/sha256sum-"$arch".txt
-    wget "$url"/rke2-images.linux-"$arch".txt
+    # Ref: https://docs.rke2.io/install/airgap
+    if [[ -n "$flags" ]] && [[ "$flags" =~ "cni" ]]; then
+      wget "$url"/rke2-images-core.linux-"$arch".txt
+      if [ -n "$tarball_type" ]; then
+        wget "$url"/rke2-images-core.linux-"$arch"."$tarball_type"
```
Comment (fmoral2) [inline modules/airgap/setup/get_artifacts.sh:46]: since we have a bunch of wgets, can we add a "download" or something, function? 

like doing the wget -O to somewhere and handling possible error

than u just call this func whenever needed on get assets and get_cni_assets

File: shared/airgap.go [hunk 7]
```diff
@@ -8,18 +8,27 @@ import (
 	"github.com/rancher/distros-test-framework/pkg/customflag"
 )
 
-// SetupPrivateRegistry sets bastion node as private registry.
-func SetupPrivateRegistry(cluster *Cluster, flags *customflag.FlagConfig) error {
+// SetupAirgapRegistry sets bastion node for airgap registry.
```
Comment (fmoral2) [inline shared/airgap.go:11]: i know we discussed at the time, but sorry...i cant remember.

why do we decided to create a airgap file here instead of letting it all on the `/pkg/airgap.go` ?

since its only going to be used there right

Comment (fmoral2) [reply shared/airgap.go:11]: cause what u did know its perfect like

having a airgap.go with the `TestBuildAirgapCluster` and aux functions together and a separated file for private and default

File: pkg/testcase/systemdefaultregistry.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+	err := shared.SetupAirgapRegistry(cluster, flags, SystemDefaultRegistry)
+	Expect(err).To(BeNil(), err)
+
+	shared.LogLevel("info", "Copying assets on the airgap nodes...")
+	err = shared.CopyAssetsOnNodes(cluster, SystemDefaultRegistry)
+	Expect(err).To(BeNil(), err)
+
+	shared.LogLevel("info", "Installing %v on airgap nodes...", cluster.Config.Product)
+	installOnServers(cluster)
+	installOnAgents(cluster)
```
[the remark this answers is not in the document]
Comment (fmoral2) [reply pkg/testcase/systemdefaultregistry.go:21]: *nit pick

File: modules/airgap/setup/get_artifacts.sh (hunk 6 above)
Comment (fmoral2) [reply modules/airgap/setup/get_artifacts.sh:46]: *design

File: entrypoint/airgap/systemdefaultregistry_test.go (hunk 1 above)
Comment (mdrahman-suse) [reply entrypoint/airgap/systemdefaultregistry_test.go:43]: I want to be able to deploy workload and validate cluster-ip, pod-to-pod communication etc... nothing outlined as of now

File: modules/airgap/setup/get_artifacts.sh (hunk 3 above)
Comment (mdrahman-suse) [reply modules/airgap/setup/get_artifacts.sh:43]: flags here referring to `server_flags`, I can rename it to be clear

File: pkg/testcase/systemdefaultregistry.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (mdrahman-suse) [reply pkg/testcase/systemdefaultregistry.go:21]: sure can add that

File: modules/airgap/setup/get_artifacts.sh (hunk 6 above)
Comment (mdrahman-suse) [reply modules/airgap/setup/get_artifacts.sh:46]: Yeah I was thinking along the same line like adding a retry logic in case it fails to download...

File: shared/airgap.go (hunk 7 above)
Comment (mdrahman-suse) [reply shared/airgap.go:11]: > i know we discussed at the time, but sorry...i cant remember.
> 
> why do we decided to create a airgap file here instead of letting it all on the `/pkg/airgap.go` ?
> 
> since its only going to be used there right

No worries, I can explain:
- `pkg/testcase/airgap.go` holds the shared **test** func that can be used by other airgap tests like `privateregistry` or `systemdefaultregistry`
- `shared/airgap.go` holds the supported func used by `pkg/testcase/airgap.go` and those func are not tests but more like func needed for the test...

I am thinking should I move `shared/airgap.go` to `support/airgap.go` to live closer to the test but then I was thinking that `shared/airgap.go` uses a lot of stuff from files in shared dir... so trying to decide that

File: entrypoint/airgap/systemdefaultregistry_test.go (hunk 1 above)
Comment (fmoral2) [reply entrypoint/airgap/systemdefaultregistry_test.go:43]: ow ok, got it

File: shared/airgap.go (hunk 7 above)
Comment (fmoral2) [reply shared/airgap.go:11]: ok tks

File: modules/airgap/setup/get_artifacts.sh (hunk 6 above)
Comment (mdrahman-suse) [reply modules/airgap/setup/get_artifacts.sh]: Updated please check

File: modules/airgap/setup/get_artifacts.sh (hunk 5 above)
Comment (mdrahman-suse) [reply modules/airgap/setup/get_artifacts.sh]: Updated

File: modules/airgap/setup/get_artifacts.sh (hunk 4 above)
Comment (mdrahman-suse) [reply modules/airgap/setup/get_artifacts.sh]: Updated

File: pkg/testcase/systemdefaultregistry.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (mdrahman-suse) [reply pkg/testcase/systemdefaultregistry.go:22]: Updated

File: modules/airgap/setup/get_artifacts.sh (hunk 3 above)
Comment (mdrahman-suse) [reply modules/airgap/setup/get_artifacts.sh]: Updated

File: entrypoint/airgap/airgap_suite_test.go (hunk 2 above)
Comment (mdrahman-suse) [reply entrypoint/airgap/airgap_suite_test.go]: Updated

File: modules/airgap/setup/get_artifacts.sh (hunk 5 above)
Comment (fmoral2) [reply modules/airgap/setup/get_artifacts.sh]: not seeing

File: modules/airgap/setup/get_artifacts.sh [hunk 9]
```diff
@@ -26,43 +46,79 @@ check_arch(){
   fi
 }
 
+download_retry() {
+  cmd=${1}
+  max_attempts=3
+  attempt_num=1
+
+  while [ $attempt_num -le $max_attempts ]; do
+
```
Comment (fmoral2) [inline modules/airgap/setup/get_artifacts.sh:55]: extra line

**Review by fmoral2 — DISMISSED** [review]

**Review by ShylajaDevadiga — APPROVED** [review]

File: modules/airgap/setup/get_artifacts.sh (hunk 9 above)
Comment (mdrahman-suse) [reply modules/airgap/setup/get_artifacts.sh]: updated

**Review by fmoral2 — APPROVED** [review]

**Review by VestigeJ — APPROVED** [review]
